pcsx2/GS/Renderers/HW/GSRendererHW.cpp

```diff
@@ -2774,62 +3094,244 @@ void GSRendererHW::Draw()
+			}
 		}
+		else if (rt->m_TEX0.TBP0 != m_cached_ctx.FRAME.Block())
+		{
+			int vertical_offset = ((static_cast<int>(m_cached_ctx.FRAME.Block() - rt->m_TEX0.TBP0) / 32) / std::max(static_cast<int>(rt->m_TEX0.TBW), 1)) * frame_psm.pgs.y; // I know I could just not shift it..
+			int texture_offset = 0;
+			int horizontal_offset = ((static_cast<int>((m_cached_ctx.FRAME.Block() - rt->m_TEX0.TBP0)) / 32) % static_cast<int>(std::max(rt->m_TEX0.TBW, 1U))) * frame_psm.pgs.x;
+			// Used to reduce the offset made later in channel shuffles
+			m_target_offset = std::abs(static_cast<int>((m_cached_ctx.FRAME.Block() - rt->m_TEX0.TBP0)) >> 5);
+
+			if (vertical_offset < 0)
+			{
+				rt->m_TEX0.TBP0 = m_cached_ctx.FRAME.Block();
+				GSVector2i new_size = rt->m_unscaled_size;
+				// Make sure to use the original format for the offset.
+				const int new_offset = std::abs((vertical_offset / frame_psm.pgs.y) * GSLocalMemory::m_psm[rt->m_TEX0.PSM].pgs.y);
+				texture_offset = new_offset;
+
+				new_size.y += new_offset;
+
+				rt->ResizeTexture(new_size.x, new_size.y, true, true, GSVector4i::loadh(new_size * rt->m_scale).loadl(GSVector2i(0, new_offset * rt->m_scale)));
+
+				if (src && src->m_from_target && src->m_from_target == rt && src->m_target_direct)
+				{
+					src->m_texture = rt->m_texture;
+				}
+
+				rt->m_valid.y += new_offset;
+				rt->m_valid.w += new_offset;
+				rt->m_drawn_since_read.y += new_offset;
+				rt->m_drawn_since_read.w += new_offset;
+
+				if (rt->m_dirty.size())
+				{
+					for (int i = 0; i < rt->m_dirty.size(); i++)
+					{
+						rt->m_dirty[i].r.y += new_offset;
+						rt->m_dirty[i].r.w += new_offset;
+					}
+				}
+
+				t_size.y += std::abs(vertical_offset);
+				vertical_offset = 0;
+			}
+
+			if (horizontal_offset < 0)
+			{
+				// Thankfully this doesn't really happen, but catwoman moves the framebuffer backwards 1 page with a channel shuffle, which is really messy and not easy to deal with.
+				// Hopefully the quick channel shuffle will just guess this and run with it.
+				rt->m_TEX0.TBP0 += horizontal_offset;
+				horizontal_offset = 0;
+			}
+
+			if (vertical_offset || horizontal_offset)
+			{
+				// Z isn't offset but RT is, so we need a temp Z to align it, hopefully nothing will ever write to the Z too, right??
+				if (ds && vertical_offset && (m_cached_ctx.ZBUF.Block() - ds->m_TEX0.TBP0) != (m_cached_ctx.FRAME.Block() - rt->m_TEX0.TBP0))
+				{
+					const int z_vertical_offset = ((static_cast<int>(m_cached_ctx.ZBUF.Block() - ds->m_TEX0.TBP0) / 32) / std::max(rt->m_TEX0.TBW, 1U)) * GSLocalMemory::m_psm[m_cached_ctx.ZBUF.PSM].pgs.y;
+					if (g_texture_cache->GetTemporaryZ() != nullptr)
+					{
+						GSTextureCache::TempZAddress z_address_info = g_texture_cache->GetTemporaryZInfo();
+
+						if (ds->m_TEX0.TBP0 != z_address_info.ZBP || z_address_info.offset != (vertical_offset - z_vertical_offset))
```

Comment:
Complaining for these ones too `z_address_info.offset != (vertical_offset - z_vertical_offset)` u32 vs int.